### PR TITLE
Adding support for telescope containers and ties in astrology.lic

### PR DESCRIPTION
**Pull request by jayRyan24**
I will add support for walkingastro as well. 

Settings:

```
telescope_storage:
  tied: toolstrap
```
or 
```
telescope_storage:
  container: backpack
```

If both are listed, it will tie before using the container. 

Changes for #3763.

### CODE REVIEW

**Comment by jayRyan24** [comment]
https://pastebin.com/CXRJPVsN

File: profiles/base.yaml [hunk 1]
```diff
@@ -1120,6 +1120,7 @@ held_athletics_items:
 # https://elanthipedia.play.net/Lich_script_repository#astrology
 astrology_buffs:
 have_telescope: false
+telescope_storage:
```
Comment (jayRyan24) [inline profiles/base.yaml]: Do I need this if I also have it in base-empty, I can't remember.

**Comment by Kaesken** [comment]
I updated everything and added the tied as stated above but it keeps wanting to use "get my telescope" instead of untying it from my toolbelt.

File: common-moonmage.lic [hunk 2]
```diff
@@ -12,8 +12,24 @@ module DRCMM
     DRC.bput('predict weather', 'You predict that', 'You are far too', 'you lack the skill to grasp them fully', 'roundtime')
   end
 
-  def get_telescope
-    DRC.bput('get my telescope', 'you get', 'You are already holding that.', "That can't be picked up", 'You need a free hand to pick that up.', 'You should stop practicing your Athletics skill before you do that')
+  def get_telescope(storage)
```
Comment (rpherbig) [inline common-moonmage.lic]: ❌  Could you pull the `common-moonmage` changes into a separate PR? We want to merge it before the other PR to give people time to get the update.

File: profiles/base.yaml (hunk 1 above)
Comment (rpherbig) [reply profiles/base.yaml]: Yes

File: profiles/base-empty.yaml [hunk 3]
```diff
@@ -67,3 +67,4 @@ empty_values:
   held_athletics_items: []
   training_list: []
   stabbity: {}
+  telescope_storage: []
```
Comment (rpherbig) [inline profiles/base-empty.yaml]: ❓ Shouldn't this be a hash? `telescope_storage: {}`

Comment (jayRyan24) [reply profiles/base-empty.yaml]: Yes

**Comment by jayRyan24** [comment]
> I updated everything and added the tied as stated above but it keeps wanting to use "get my telescope" instead of untying it from my toolbelt.

This is probably in reference to what @rpherbig is talking about. You would need to pull in the common-moonmage changes, and then run ;common-moonmage before any of the other stuff will work.

File: common-moonmage.lic (hunk 2 above)
Comment (jayRyan24) [reply common-moonmage.lic]: Yeah, will do that.

**Comment by rpherbig** [comment]
✔️ Looks good. If I haven't merged this in a few days, remind me.

**Comment by jayRyan24** [comment]
> ✔️ Looks good. If I haven't merged this in a few days, remind me.

Will do.

**Comment by jayRyan24** [comment]
@rpherbig I did add changes to the file just now. Basically, Astrology will constantly get & stow the telescope each iteration. I moved that logic out (now that it's methodized), before and after the observe logic.

File: astrology.lic [hunk 4]
```diff
@@ -168,7 +169,7 @@ class Astrology
     cast_spell(@rtr_data, settings)
     perc_time = Time.now - 61
     Flags.add('rtr-expire', get_data('spells').spell_data['Read the Ripples']['expire']) if rtr_active?
-    DRCMM.get_telescope(@tele_storage) if @have_telescope
+    get_telescope(@tele_storage) if @have_telescope
```
Comment (rpherbig) [inline astrology.lic]: 💡 I like not having the prefix so the code is concise, but I've recently decided that, overall, I prefer to have it. It makes it easier for new contributors to figure out where helpers come from.

Not required for you to change this, just getting my thoughts out there.

Comment (jayRyan24) [reply astrology.lic]: My plan is to make DRCMM's methods more robust and add them to astrology.lic. 

```
if @have_telescope
      case center_telescope_on(target)
      when /open it/
```

Comment (jayRyan24) [reply astrology.lic]: I will do whatever you want me to do in this situation, just let me know. :)

Comment (jayRyan24) [reply astrology.lic]:
```
  def roll_bones(is_tied, container)
    if get_bones
      DRC.bput('roll my bones', 'roundtime')
      store_bones
    end
  end
```

Comment (rpherbig) [reply astrology.lic]: I love the plan.

As for the prefix, if you come back to this file for some reason please add it, no need to bother with another commit just for that change.

Comment (jayRyan24) [reply astrology.lic]: I'm just waiting for all of these things to get merged, then I will update the whole.

**Comment by Kaesken** [comment]
Maybe I'm not understanding?  I replaced each of the files with what's in the file change tab.

Log:  https://pastebin.com/zAqG5P8V
YAML:  https://pastebin.com/P9KtnWZw

**Comment by jayRyan24** [comment]
> Maybe I'm not understanding? I replaced each of the files with what's in the file change tab.
> 
> Log: https://pastebin.com/zAqG5P8V
> YAML: https://pastebin.com/P9KtnWZw
> 
> Even tried killing common-moonmage and restarting it but it won't tell me whether or not it's been started and wont' do anything for it so I run training-manager. How can I get common-moonmage to work by itself? I apologize in advance for any difficulty on my part, lol. This section of github is really new to me.

Hey, sorry you're having so much trouble. So basically, you would need to start up the system and let everything load

`,kill walkingastro`

Due to how the file system works, whenever you run dependency, it will update all the files and bring in new changes. This can be stopped with developer mode, but it's not something you should do unless you know how to handle it. What you can do is run the system like normal, and then make all the changes to the files with copy and paste. 

Then:
```
,common-moonmage
,walkingastro
```

This should then start up and let you test. But once you restart dependency, i.e. log out and back in, or even `,k dependency` and `,dependency`, it will overwrite whatever changes you had made.

**Comment by Kaesken** [comment]
Oh, no need to be sorry, hun!  I am in developer mode because I do edit a few of the files but they're just a few tweaks.  I can get off dev mode and have dependency update everything then go back and remake my changed and see if it'll work.  I didn't even think of that.  I didn't restart dependency, tho.  Going to try all that now.  :)

File: astrology.lic (hunk 4 above)
Comment (jayRyan24) [reply astrology.lic]: I reverted the changes as I would just be doing that in the next PR. No worries!

**Comment by jayRyan24** [comment]
You'll lose your changes. If in dev mode, you should already be fine. You would just need to copy and paste all those files. But it did go through, so you'll get the changes there I guess, lol.

**Comment by jayRyan24** [comment]
@rpherbig Here's the log with the get/stow changes. A lot more clean as we aren't getting and stowing each observe. 

https://pastebin.com/qi6r0V6k

**Comment by Kaesken** [comment]
walkingastro wasn't running, but then after I killed dependency and restarted it, this is what I got when starting walkingastro:  https://pastebin.com/6AzCe4Gf

This is what I get when trying to stop/start common-moonmage:  https://pastebin.com/qSzVVfi9
I can't even tell if it does anything or what it does.

I also forgot to mention that I was trying to run "training-manager" with the telescope.  I have it in settings for her to do astro in her hunting which isn't untying it which it still tries to "get my telescope" in the combat-trainer.

[combat-trainer]>get my telescope
You should untie the e'erdream telescope from the teller's toolbelt first.

I completely forgot to check if that was possible or if you guys were just working on the other files first.  My apologies on that.

But!  The walking astro one works!  Just the match doesn't match for me, hehe

**Comment by Kaesken** [comment]
Ok, so I got my walkingastro to finally work with untying the telescope!!

I had to add 'You untie' to this part under walkingastro.lic
  def get_telescope
    case DRCMM.get_telescope(@tele_storage)
    when 'You get', 'You remove', 'You untie'

And change 'you untie your' to 'You untie' in common-moonmage and it's working perfectly for walkingastro.

**Comment by jayRyan24** [comment]
OK, thanks a lot. I didn't have a tie to test with!

**Comment by jayRyan24** [comment]
@rpherbig Can we get this merged? I can handle the questions (and enough people are using the change) that we can just instruct people to restart dep to pick up the changes. 

I am kind of in the mood to get the changes for the common file and astrology moving further.

**Comment by Tarhish** [comment]
The only reason I'm aware of is that some people run multiple characters out of the same lich setup.  Certain common files don't get updated if they're being used at the time you login, so there have been issues in the past with everything crashing and burning because someone got one script updated but not the common file it relies on.

That said, this is common-moonmage.lic, not common.lic so I'd guess chances are low that it's going to cause problems.

**Comment by jayRyan24** [comment]
Yup, that's it! Multiple chars! Anywho, I would like to just move this in at this point! If not, it's fine, I'm just bored!

**Comment by rpherbig** [comment]
I'll merge it tomorrow

**Comment by Tarhish** [comment]
Since I'm poking around anyway yeah, this and the common change look like it should work out to me, so thanks for adding it!

I have a tiny suggestion: long term I think the base.yaml entry for telescope_storage: should look like divination_bones_storage:
```
divination_bones_storage:
  container:
  tied:
```
I don't think it has any technical impact whatsoever, but it'll let someone looking in base know that's what it's supposed to look like.

**Comment by jayRyan24** [comment]
I saw it but didn't want to add both and make them think they'd have to have them both set. So I left them out. I did add to the repo info on the wiki though to explain the setting. And they will also be able to check the YAMLs for MM's inside the repo.

**Pull request merged**